src/sympc/module/nn/functional.py

```diff
@@ -37,3 +49,323 @@ def mse_loss(pred: MPCTensor, target: MPCTensor, reduction: str = "mean") -> MPC
+        kernel_size (Tuple[int, int]): the kernel size
+        stride (Tuple[int, int]): the stride size
+        padding (Tuple[int, int]): the padding size
+        dilation (Tuple[int, int]): the dilation size
+
+    Returns:
+        The prepared share tensor (reshaped)
+    """
+    session = get_session(x.session_uuid)
+    tensor = x.tensor.numpy()
+
+    padding = [(0, 0)] * len(tensor.shape[:-2]) + [
+        (padding[0], padding[0]),
+        (padding[1], padding[1]),
+    ]
+    tensor_type = session.tensor_type
+
+    padding_value = 0
+    if session.rank == 0:
+        # ATTENTION: Min value for max_pool2d that works -2**8
+        # This is because 2 is the base and 16 is the precision 2 ** 24 / (2 ** 16) = 2 ** 9
+        padding_value = -(2 ** 24)
+
+    tensor = np.pad(tensor, padding, mode="constant", constant_values=padding_value)
+
+    output_shape = tensor.shape[:-2]
+    output_shape += (
+        (tensor.shape[-2] - kernel_size[0]) // stride[0] + 1,
+        (tensor.shape[-1] - kernel_size[1]) // stride[1] + 1,
+    )
+    output_shape += kernel_size
+
+    output_strides = tensor.strides[:-2]
+    output_strides += (stride[0] * tensor.strides[-2], stride[1] * tensor.strides[-1])
+    output_strides += tensor.strides[-2:]
+
+    window_view_share = torch.tensor(
+        np.lib.stride_tricks.as_strided(
+            tensor, shape=output_shape, strides=output_strides
+        ),
+        dtype=tensor_type,
+    )
+
+    window_view_share = window_view_share.reshape(-1, *kernel_size)
+
+    res_share = ShareTensor(config=x.config)
+    res_share.tensor = window_view_share
+    return res_share
+
+
+def max_pool2d(
+    x: MPCTensor,
+    kernel_size: Union[int, Tuple[int, int]],
+    stride: Optional[Union[int, Tuple[int, int]]] = None,
+    padding: Union[int, Tuple[int, int]] = 0,
+    dilation: Union[int, Tuple[int, int]] = 1,
+) -> MPCTensor:
+    """Compute the max poold for a tensor with 2 dimension.
+
+    Args:
+        x (MPCTensor): the MPCTensor on which to apply the operation
+        kernel_size (Union[int, Tuple[int, int]]): the kernel size
+            in case it is passed as an integer then that specific value is used for height and width
+        stride (Union[int, Tuple[int, int]]): the stride size
+            in case it is passed as an integer then that specific value is used for height and width
+        padding (Union[int, Tuple[int, int]]): the padding size
+            in case it is passed as an integer then that specific value is used for height and width
+        dilation (Union[int, Tuple[int, int]]): the dilation size
+            in case it is passed as an integer then that specific value is used for height and width
+
+    Returns:
+        A tuple representing maximum values and the indices (as a one hot encoding
+
+    Raises:
+        ValueError: in case kernel_size is passed as a tuple with a number of elements != 2
+    """
+    if isinstance(kernel_size, int):
```

Comment:
Maybe: A separate function for sanity checks? Lines 174 to 205. If backward and forward have the same sanity checks, you can combine them into one function and have the rest of them in the function. Or have two separate sanity check functions for backward and forward depending on the overlap. Just to keep the code clean and easier to understand.

Reply:
Yes, I agree